SharkCage/CageManager/CageDesktop.h

```diff
@@ -0,0 +1,86 @@
+#include <Windows.h>
+#include <iostream>
+
+#include "CageLabeler.h"
+#include "FullWorkArea.h"
+
+class CageDesktop
+{
+public:
+	CageDesktop(PSECURITY_DESCRIPTOR p_sd);
+	~CageDesktop();
+	bool Init();
+private:
+	HDESK old_desktop;
+	HDESK new_desktop;
+	FullWorkArea full_work_area;
+	CageLabeler cage_labeler;
+};
+
+CageDesktop::CageDesktop(PSECURITY_DESCRIPTOR p_sd)
+{
+	old_desktop = ::GetThreadDesktop(GetCurrentThreadId());
+
+	ACCESS_MASK desk_access_mask = DESKTOP_CREATEMENU
+		| DESKTOP_CREATEWINDOW
+		| DESKTOP_ENUMERATE
+		| DESKTOP_HOOKCONTROL
+		| DESKTOP_JOURNALPLAYBACK
+		| DESKTOP_JOURNALRECORD
+		| DESKTOP_READOBJECTS
+		| DESKTOP_SWITCHDESKTOP
+		| DESKTOP_WRITEOBJECTS
+		| READ_CONTROL
+		| WRITE_DAC
+		| WRITE_OWNER;
+
+	SECURITY_ATTRIBUTES sa;
+	sa.nLength = sizeof(SECURITY_ATTRIBUTES);
+	sa.lpSecurityDescriptor = p_sd;
+	sa.bInheritHandle = FALSE;
+
+	new_desktop = ::CreateDesktop(TEXT("shark_cage_desktop"), NULL, NULL, NULL, desk_access_mask, &sa);
+}
+
+CageDesktop::~CageDesktop()
+{
+	if (!::SetThreadDesktop(old_desktop))
+	{
+		std::cout << "Failed to set thread desktop back to old desktop. Error " << ::GetLastError() << std::endl;
+	}
+
+	if (!::SwitchDesktop(old_desktop))
+	{
+		std::cout << "Failed to switch back to old desktop. Error " << ::GetLastError() << std::endl;
+	}
+}
+
+
+bool CageDesktop::Init()
+{
+	if (!::SwitchDesktop(new_desktop))
+	{
+		std::cout << "Failed to switch to cage desktop. Error " << ::GetLastError() << std::endl;
+		return false;
+	}
+
+	if (!::SetThreadDesktop(new_desktop))
```

Comment:
still no switching back to the old desktop if this function fails. is this intentional?